src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultRazorComponentSearchEngine.cs

```diff
@@ -0,0 +1,83 @@
+        {
+            _foregroundDispatcher = dispatcher ?? throw new ArgumentNullException(nameof(dispatcher));
+        }
+
+        /// <summary>Search for a component in a project based on its tag name and fully qualified name.</summary>
+        /// <remarks>
+        /// This method makes several assumptions about the nature of components. First, it assumes that a component
+        /// a given name `Name` will be located in a file `Name.razor`. Second, it assumes that the namespace the
+        /// component is present in has the same name as the assembly its corresponding tag helper is loaded from.
+        /// Implicitly, this method inherits any assumptions made by TrySplitNamespaceAndType.
+        /// </remarks>
+        /// <param name="project">A project to search for the component in.</param>
+        /// <param name="componentName">The name of the component.</param>
+        /// <param name="fullyQualifiedComponentName">The fully qualified name of the component.</param>
+        /// <param name="cancellationToken">A cancellation token, primarily for the document resolver.</param>
+        /// <returns>A tuple containing the document URI and related resources loaded during search.</returns>
+        public async override Task<Tuple<Uri, DocumentSnapshot, RazorCodeDocument>> TryLocateComponent(ProjectSnapshot project, string componentName, string fullyQualifiedComponentName, CancellationToken cancellationToken)
+        {
+            foreach (var path in project.DocumentFilePaths)
+            {
+                // Rule out if not Razor component with correct name
+                if (!IsPathCandidateForComponent(path, componentName))
+                {
+                    continue;
+                }
+
+                // Get document and code document
+                var documentSnapshot = await Task.Factory.StartNew(() => project.GetDocument(path), cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+                var razorCodeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+
+                // Make sure we have the right namespace of the fully qualified name
+                if (!ComponentNamespaceMatchesFullyQualifiedName(razorCodeDocument, fullyQualifiedComponentName))
+                {
+                    continue;
+                }
+
+                var uri = new UriBuilder
+                {
+                    Scheme = "file",
+                    Path = path,
+                    Host = string.Empty,
+                }.Uri;
+                return Tuple.Create(uri, documentSnapshot, razorCodeDocument);
+            }
+            return null;
+        }
+
+        public static bool IsPathCandidateForComponent(string path, string componentName)
+        {
+            return Path.GetFileName(path).Equals($"{componentName}.razor", StringComparison.Ordinal);
```

Comment:
Doing an `EndsWith` will be more performant because you don't need to re-allocate a string for the path

Reply:
To keep the logic here wouldn't it be `EndsWith(RazorConst)` and `StartsWith(componentName)`?

Reply:
As in `path.EndsWith`? I would still need to build the `".razor"` bit. Can you elaborate?

Reply:
path.EndsWith($"{componentName}.razor, FilePathComparer.Instance)

Reply:
Ahhh, @NTaylorMullen was suggesting we avoid the allocation from `Path.GetFileName`, not the allocation from `${componentName}.razor`. Ignore me, his is better.

Reply:
@NTaylorMullen